fw/http_msg.c

```diff
@@ -815,6 +820,47 @@ __hdr_sub(TfwHttpMsg *hm, const TfwStr *hdr, unsigned int hid)
 	return TFW_PASS;
 }
 
+/**
+ * Substitute request method
+ *
+ * For a while it's just a custom realization for substituting dest method with sub method
+ * (Implemented to substitute PURGE with GET in X-Tempesta-Cache processing)
+ */
+int
+tfw_http_req_sub(TfwHttpMsg *hm, unsigned char dest_meth, unsigned char sub_meth)
+{
+	const char *dest_name = NULL;
+	const char *sub_name = NULL;
+
+	size_t dest_len, sub_len;
+
+	TfwStr sub_str = {};
+
+	if (hm->req->method != dest_meth)
+		return TFW_BLOCK; /* todo: check return values */
+
+	/* do we need to look for "get" string in HTTP headers or just set it as stub constant value? */
+	dest_name = tfw_http_tbl_method_get_name(dest_meth);
+	sub_name = tfw_http_tbl_method_get_name(sub_meth);
+
+	/* todo: optimize this later */
+	if (!dest_name || !sub_name)
+		return TFW_BLOCK;
+
+	dest_len = strlen(dest_name);
+	sub_len = strlen(sub_name);
+
+	if (dest_len<sub_len)
+		return TFW_BLOCK;
+
+	sub_str = TFW_STR_FROM_CSTR(sub_name);
+
+	if (ss_skb_cutoff_data(hm->msg.skb_head, &sub_str, 0, dest_len-sub_len))
```

Comment:
This is the place where the things are getting complicated. We receive data in socket buffers, `skb`, which may deliver HTTP messages in many chunks (imagine that `GE`, `T / H` and `TTP/1.1` are received in 3 different TCP segments). We do not copy the input data. Instead, `TfwStr` gathers all the chunks *and* keep references to the original skbs. This way you can work with fragmented data using TfwStr API, but if you need you can go down and adjust the underlying skb data and resend it.

This all means that you need the original skbs under `sub_str` to be able to call `ss_skb_cutoff_data()`